### ISSUE TITLE
FFI and error handling

### ISSUE THREAD

**Issue by vmx** [filecoin-project/rust-fil-proofs#889]
**Update**: When I was opening this issue, GitHub was proposing me similar issues and I found https://github.com/filecoin-project/rust-fil-proofs/issues/195 which explains a lot why things are as they are. I will take that into account for my future work.

----

This is more a train of thoughts than actionable items. It is what the IPLD Team calls [Exploration Report](https://github.com/ipld/specs/blob/d60f5682b0a8deddda59202cfd5d6f55b0103d4d/design/history/exploration-reports/README.md). I wanted to post this somewhere and opening an issue here seems to be the most sensible thing.

Based on my findings I will then figure out next steps. A lot of the "unified errors" stuff would've been needed if I would've used macros for catching panics. But as I will now use closures for it (see below), it is no longer a major concerns but more like a nice to have we might want to consider in order to have an easier to maintain code base.

### Intro

Most of the rust-fil-proofs stuff used to live in a single repository. Things seem to have been pretty entwined. I currently look into making the FFI more uniform so that it's:

 - Easy for the Rust folks to expose new APIs
 - Easy for the FFI consumers to use those APIs

One big there is error handling. You want to return the same kind of error, no matter from which crate it came from. The FFI consumers should have a unified way to deal with those.

#### Custom structs

Every FFI call returns its own custom struct. Though all structs have in common that they have a `status_code` and a `error_msg` field. A minor thing is that `status_code` current has a different type between different FFIs (that's captured [here](https://github.com/filecoin-project/rust-fil-proofs-ffi/issues/8), that can be fixed easily, there should be a single enum to capture all status codes throughout Filecoin. That would probably live in [rust-fil-ffi-toolkit](https://github.com/filecoin-project/rust-fil-ffi-toolkit).

#### Converting errors to FFI

In order to convert errors into something the structs returned by the FFI can use, there's kind of a "catch-all" function named [err_code_and_msg](https://github.com/filecoin-project/rust-fil-sector-builder/blob/fa155cdd33be0843f936efbe8f888bdf44d7dd8c/sector-builder-ffi/src/responses.rs#L80) that does that conversion. In there the `err.downcast_ref()` calls looked weird to me. It to me a while do understand why this is needed (more on that on the "problems" section).

#### Using failure crate in mixed mode

I haven't used the failure crate before so please correct me if I'm wrong. We are using [custom fail types](https://boats.gitlab.io/failure/custom-fail.html) which makes sense after reading the documentation about them. *But* we are also using the [Error type](https://boats.gitlab.io/failure/use-error.html) which I think is the reason why we need the `downcast_ref()` calls when converting them to FFI compatible errors. I named this approach "mixed mode" and I think we should just use the "custom fail types" instead.

Using the "Error type" has another bad side effect. You can't implement custom conversion for it (at least I haven't found a way to do it). If custom conversion would be possible, we wouldn't need to have that `err_code_and_msg()` function, but instead it would look like that (still very similar, but cleaner):

```rust
impl From<SectorBuilderErr> for FFIError {
    fn from(error: SectorBuilderErr) -> FFIError {
        let msg = CString::new(format!("{}", error)).unwrap();
        let ptr = msg.as_ptr();
        mem::forget(msg);

        let (code, message) = match error {
            SectorBuilderErr::OverflowError { .. } => (FCPResponseStatus::FCPCallerError, ptr),
            SectorBuilderErr::IncompleteWriteError { .. } => (FCPResponseStatus::FCPReceiverError, ptr),
            SectorBuilderErr::Unrecoverable(_, _) => (FCPResponseStatus::FCPReceiverError, ptr),
            SectorBuilderErr::PieceNotFound(_) => (FCPResponseStatus::FCPCallerError, ptr),
        };

        FFIError::new(code, message)
    }
}
```

This isn't currently possible, hence I'm not 100% sure if it works that way, but I think it should. It could then be used like that:

```rust
let FFIError { code, message } = FFIError::from(err);
response.status_code = code;
response.error_msg = message;
```

Where `FFIError` is defined centrally in e.g. the ffi-toolkit. I have the idea of doing such a conversion from [Mozilla's ffi_support crate](https://docs.rs/ffi-support/0.3.5/ffi_support/). See especially the section about the [ExternError](https://docs.rs/ffi-support/0.3.5/ffi_support/struct.ExternError.html).

When I started the work on the FFI with the "catching panics in the FFI on the Rust side" PR, I thought I would be using macros for it. There I would've needed such a general conversion, as you can't easily call out to "random" functions. Though, now I want to use [closures just as ffi_support is doing it](https://docs.rs/ffi-support/0.3.5/ffi_support/fn.call_with_result.html). This way we can do within those closures whatever we like, even very module specific things.

To conclude: I think it makes sense to use custom fail types only.

#### Having the error as part of the response code

Currently error and success is put into a single struct. A lot of manual construction of those structs is happening on the Rust side. If the error would be separate, you could do automatic conversion like ffi_support is doing it (or even use ffi_support). Please note that I don't want to change that soon. I want to improve things incrementally and not breaking the world. I just thought I bring it up as it is something we might want to keep in mind.

**Comment by laser** [comment]
@vmx Thank you for this write-up. I have some questions:

> If the error would be separate, you could do automatic conversion like ffi_support is doing it (or even use ffi_support).

I don't quite understand what you are proposing. Would you mind giving an example of how you would rewrite one of our current functions to use this alternative approach? Sketching out the revised types would be sufficient, I bet.

> I named this approach "mixed mode" and I think we should just use the "custom fail types" instead.

I believe that we should be using custom fail types as well. 

We are going to need to provide information to the caller which explains the nature of the error which they received, e.g. whether its the receiver's fault (the disk is full) or the sender's fault (malformed sector id). I am fairly certain that we operate in "mixed mode" because we got lazy about coming up with different names for the various failure modes in our system. 

I also think it's not particularly straightforward to map (at the system's edge) from something like `FileNotFoundError` to a "caller screwed up" or "callee screwed up" error code, as the originator of the error message may be dozens of stack frames from the place where we need to perform that mapping. If you have some ideas for how to do this in a sane manner, please do share.

> there should be a single enum to capture all status codes throughout Filecoin. That would probably live in rust-fil-ffi-toolkit.

This is a wonderful suggestion.

**Comment by vmx** [comment]
> I don't quite understand what you are proposing. Would you mind giving an example of how you would rewrite one of our current functions to use this alternative approach? Sketching out the revised types would be sufficient, I bet.

If you look at the [example from ffi_supports `call_with result()`](https://docs.rs/ffi-support/0.3.5/ffi_support/fn.call_with_result.html):

```rust
#[no_mangle]
pub extern "C" fn mylib_print_string(
    thing_to_print: FfiStr<'_>,
    error: &mut ExternError
) {
    ffi_support::call_with_result(error, || {
        let s = thing_to_print.as_str();
        if s.is_empty() {
            return Err(BadEmptyString);
        }
        println!("{}", s);
        Ok(())
    })
}
```

In case of an error you just return a normal, just native error `Err(BadEmptyString)`. But the wraper function `ffi_support::call_with_result()`, will make sure it gets properly converted into an error the other side of the FFI can easily consume.



> I believe that we should be using custom fail types as well.
> […]

Thanks for the infos. I need to put more thought into that.

**Comment by vmx** [comment]
I've spend my day reading a lot about Rust error handling and `failure`. I concentrate on making it possible to have the errors kind of at a central location, similar to the way Mozilla's ffi-support is doing it. I concluded that it's not possible, as we are using `failure`. I then wondered how ffi-support is actually doing it. Then I saw that they are also using `failure`. So there must be a way. I'll dig into that a bit more next time.
